### PR TITLE
Refactor ImageTextDataset to its own file. Implement error handling a…

### PR DESCRIPTION
**Pull request by afiaka87**
…nd index skipping in ImageTextDataset. Refactor args handling in train_dalle.py.

### CODE REVIEW

**Comment by afiaka87** [comment]
I also made `--truncate_captions` have the `store_true` property so that you can type `--truncate_captions` instead of `--truncate_captions True`

**Comment by afiaka87** [comment]
@janEbert Always appreciate some extra code review if you aren't busy.

File: train_dalle.py [hunk 1]
```diff
@@ -184,102 +181,59 @@ def group_weight(model):
...
-        description = choice(descriptions)
-
-        tokenized_text = tokenizer.tokenize(description, self.text_len, truncate_text=args.truncate_captions).squeeze(0)
-
-        image_tensor = self.image_tranform(image)
-        return tokenized_text, image_tensor
 
 # create dataset and dataloader
 
+is_shuffle = distributed_utils.using_backend(distributed_utils.HorovodBackend)
```
Comment (janEbert) [inline train_dalle.py:187]: Please negate this, we want to _avoid_ shuffling only if using Horovod. :)

File: train_dalle.py [hunk 2]
```diff
@@ -1,33 +1,22 @@
...
 
-# torch
-
 import torch
+
 from torch.nn.utils import clip_grad_norm_
 from torch.optim import Adam
 from torch.optim.lr_scheduler import ReduceLROnPlateau
+from torch.utils.data import DataLoader
+import argparse
```
Comment (janEbert) [inline train_dalle.py:10]: You can remove this `argparse` import; it's already imported (and the PEP 8 import order would be violated here as `argparse` is in the stdlib).

File: train_dalle.py [hunk 3]
```diff
@@ -211,10 +211,12 @@ def group_weight(model):
 
 # initialize DALL-E
 
-dalle = DALLE(vae=vae, **dalle_params)
-if args.fp16:
-    dalle = dalle.half()
-dalle = dalle.cuda()
+
+dalle = DALLE(vae = vae, **dalle_params)
```
Comment (janEbert) [inline train_dalle.py:215]: I thought you wanted those kwargs without spaces. ;)

File: train_dalle.py [hunk 4]
```diff
@@ -36,9 +23,9 @@
                     help='path to your partially trained DALL-E')
 
 parser.add_argument('--image_text_folder', type = str, required = True,
-                    help='path to your folder of images and text for learning the DALL-E')
+    help='path to your folder of images and text for learning the DALL-E')
 
-parser.add_argument('--truncate_captions', dest='truncate_captions',
+parser.add_argument('--truncate_captions', dest='truncate_captions', action = 'store_true',
```
[the remark this answers is not in the document]
Comment (afiaka87) [reply train_dalle.py]: well since you're in support of it ha - i went ahead and ran autopep8 on train_dalle.py. I don't really mind one way or the other so long as we're all on the same page.

File: train_dalle.py (hunk 3 above)
Comment (afiaka87) [reply train_dalle.py]: It would appear that I do, but only on code I touch ha. I just ran autopep8 on it.

**Comment by afiaka87** [comment]
okay @lucidrains this one's ready to go when you are.

File: .gitignore [hunk 5]
```diff
@@ -1,5 +1,8 @@
 # dall-e generation outputs
 outputs/
+dalle.pt
```
Comment (lucidrains) [inline .gitignore:3]: hmm, perhaps this should be *.pt ?

File: train_dalle.py [hunk 6]
```diff
@@ -323,14 +277,14 @@ def __getitem__(self, ind):
 avoid_model_calls = using_deepspeed and args.fp16
 
 # training
-
+torch.cuda.empty_cache() # Avoid allocation error due to potential bug in deepspeed. See https://github.com/lucidrains/DALLE-pytorch/issues/161
```
Comment (lucidrains) [inline train_dalle.py:280]: i thought @janEbert removed this in an earlier pull request

Comment (janEbert) [reply train_dalle.py:280]: Yeah, you can remove this line. :)

File: train_dalle.py [hunk 7]
```diff
@@ -212,11 +216,10 @@ def group_weight(model):
...
 
-dalle = DALLE(vae = vae, **dalle_params)
-if not using_deepspeed:
-    if args.fp16:
-        dalle = dalle.half()
-    dalle = dalle.cuda()
+dalle = DALLE(vae=vae, **dalle_params)
+if not using_deepspeed and args.fp16:
+    dalle = dalle.half()
+dalle = dalle.cuda()
```
Comment (janEbert) [inline train_dalle.py]: Here, please keep moving the model to the GPU in the `if not using_deepspeed` block. We want to let DeepSpeed handle everything after model creation, both FP16 conversion and moving to GPU.

File: train_dalle.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (janEbert) [reply train_dalle.py]: I like consistency, but it's not my codebase, so my opinion really doesn't matter. :p

File: train_dalle.py (hunk 7 above)
Comment (afiaka87) [reply train_dalle.py]: Gotcha that was a mis-merge. Good catch.

**Comment by afiaka87** [comment]
Thanks for the review @lucidrains @janEbert - going to re-run this branch real quick just to make sure everything works but I believe we caught everything.

**Comment by afiaka87** [comment]
@lucidrains alright i'm using it currently to run a hyperparameter sweep https://wandb.ai/afiaka87/sweep_dalle?workspace=user-afiaka87

something like 20 successful 1000 iteration runs - if that's not an integration test I don't know what is.

**Comment by afiaka87** [comment]
already coming in handy too:
```sh
0 180 loss - 4.784719467163086
0 190 loss - 5.3411993980407715
0 200 loss - 5.28896951675415
0 210 loss - 5.1630167961120605
An exception occurred trying to load file /root/Datasets/Previews/previews_390/3901727.jpg.
Skipping index 6465188
0 220 loss - 5.059352874755859
0 230 loss - 5.032174110412598
0 240 loss - 5.2245259284973145
0 250 loss - 5.022186279296875
0 260 loss - 5.176838397979736
0 270 loss - 4.998762130737305
0 280 loss - 5.168933391571045
0 290 loss - 4.991147518157959
0 300 loss - 4.944616794586182
0 310 loss - 4.800735950469971
0 320 loss - 5.075268745422363
0 330 loss - 4.961371421813965
0 340 loss - 4.879711151123047
0 350 loss - 5.025337219238281
0 360 loss - 5.033573150634766
An exception occurred trying to load file /root/Datasets/Previews/previews_570/5705003.jpg.
Skipping index 5254782
0 370 loss - 5.12672758102417
0 380 loss - 5.011307716369629
0 390 loss - 5.002786636352539 
```

**Comment by lucidrains** [comment]
Thanks! Lgtm!
